x/kavadist/types/params.go

```diff
@@ -90,22 +90,54 @@ func ParamKeyTable() params.KeyTable {
 // ParamSetPairs implements the ParamSet interface and returns all the key/value pairs
 func (p *Params) ParamSetPairs() params.ParamSetPairs {
 	return params.ParamSetPairs{
-		{Key: KeyActive, Value: &p.Active},
-		{Key: KeyPeriods, Value: &p.Periods},
+		params.NewParamSetPair(KeyActive, &p.Active, validateActiveParam),
+		params.NewParamSetPair(KeyPeriods, &p.Periods, validatePeriodsParams),
 	}
 }
 
 // Validate checks that the parameters have valid values.
 func (p Params) Validate() error {
+	if err := validateActiveParam(p.Active); err != nil {
+		return err
+	}
+
+	return validatePeriodsParams(p.Periods)
+}
+
+func validateActiveParam(i interface{}) error {
+	_, ok := i.(bool)
+	if !ok {
+		return fmt.Errorf("invalid parameter type: %T", i)
+	}
+
+	return nil
+}
+
+func validatePeriodsParams(i interface{}) error {
+	periods, ok := i.(Periods)
+	if !ok {
+		return fmt.Errorf("invalid parameter type: %T", i)
+	}
+
 	prevEnd := tmtime.Canonical(time.Unix(0, 0))
-	for _, pr := range p.Periods {
+	now := time.Now()
```

Comment:
I think this could cause validators to disagree on the hash of a block if they have different clock times, and a param change proposal is accepted with a period with an End time that sits in between the validators' clock times.
I'd remove the `time.Now()` call.